lib/ddtrace/profiling/ext/thread.rb

```diff
@@ -0,0 +1,72 @@
+require 'ffi'
+
+module Datadog
+  module Profiling
+    module Ext
+      # C-struct for retrieving clock ID from pthread
+      class CClockId < FFI::Struct
+        layout :value, :int
+      end
+
+      # Extensions for pthread-backed Ruby threads, to retrieve
+      # the thread ID, clock ID, and CPU time.
+      module CThread
+        extend FFI::Library
+        ffi_lib 'ruby', 'pthread'
+        attach_function :rb_nativethread_self, [], :ulong
+        attach_function :pthread_getcpuclockid, [:ulong, CClockId], :int
+
+        def self.prepended(base)
+          # Be sure to update the current thread too; as it wouldn't have been set.
+          Thread.current.send(:set_native_ids)
+        end
+
+        attr_reader \
+          :clock_id,
+          :native_thread_id
+
+        def initialize(*args)
+          # Wrap the work block with our own
+          # so we can retrieve the native thread ID within the thread's context.
+          wrapped_block = proc do |*t_args|
+            # Set native thread ID & clock ID
+            set_native_ids
+            yield(*t_args)
+          end
+
+          super(*args, &wrapped_block)
+        end
+
+        def cpu_time(unit = :float_second)
+          return unless clock_id && Process.respond_to?(:clock_gettime)
+          Process.clock_gettime(clock_id, unit)
+        end
+
+        private
+
+        def set_native_ids
+          # Only run if invoked from same thread, otherwise
+          # it will receive incorrect thread and clock IDs.
+          return unless Thread.current == self
+          @native_thread_id = get_native_thread_id
+          @clock_id = get_clock_id(@native_thread_id)
+        end
+
+        def get_native_thread_id
```

Comment:
I wouldn't use the term `native` id as it's actually more linked to `pthread` than native.
Python exposes something it calls the _native_ id and is actually a `pid_t` coming from `gettid()`: https://man7.org/linux/man-pages/man2/gettid.2.html

which might be closer than native than the `pthread` id.

Reply:
I'm using Ruby terminology here, because although right now this is `pthread` only, Ruby uses the term `native` to abstract from the platform in its C library. I think eventually, if we can, it'd be nice to expand this beyond `pthread`.